pkg/controllers/workgenerator/controller.go

```diff
@@ -779,20 +877,81 @@ func setBindingStatus(works map[string]*fleetv1beta1.Work, resourceBinding *flee
+			// might be incomplete or stale; apply/availability check failure and drifts cannot occur in
+			// report diff mode).
+		case appliedSummarizedStatus == workConditionSummarizedStatusIncomplete:
+			// The ClientSideApply or ServerSideApply apply strategy is in use but some of the works have
+			// not been applied yet.
+			//
+			// In this case, no diffed, failed, or drifted placements will be set (as information present
+			// might be incomplete or stale).
+		case appliedSummarizedStatus == workConditionSummarizedStatusFalse:
+			// The ClientSideApply or ServerSideApply apply strategy is in use but some of the works have
+			// apply op failures.
+			//
+			// In this case, set failed, diffed, and drifted placements.
 			failedManifests := extractFailedResourcePlacementsFromWork(w)
 			failedResourcePlacements = append(failedResourcePlacements, failedManifests...)
-		}
-		// Diffed placements can only appear when the Applied condition is set to False.
-		if appliedCond.Status == metav1.ConditionFalse {
+
 			diffedManifests := extractDiffedResourcePlacementsFromWork(w)
 			diffedResourcePlacements = append(diffedResourcePlacements, diffedManifests...)
+
+			driftedManifests := extractDriftedResourcePlacementsFromWork(w)
+			driftedResourcePlacements = append(driftedResourcePlacements, driftedManifests...)
+		case availabilitySummarizedStatus == workConditionSummarizedStatusIncomplete:
+			// The ClientSideApply or ServerSideApply apply strategy is in use; all works have been applied but
+			// some of the works have not completed the availability check yet.
+			//
+			// In theory this would not happen as the Fleet work applier will always set the Applied and
+			// Available conditions together. However, Fleet can still handle this case for completeness reasons.
+			//
+			// In this case, set drifted placements; no failed or diffed placements will be set (availability
+			// check failure information might be incomplete or stale; diffs will only occur when there
+			// is an apply failure or the report diff mode is on).
+			driftedManifests := extractDriftedResourcePlacementsFromWork(w)
+			driftedResourcePlacements = append(driftedResourcePlacements, driftedManifests...)
+		case availabilitySummarizedStatus == workConditionSummarizedStatusFalse:
+			// The ClientSideApply or ServerSideApply apply strategy is in use; all works have been applied but
+			// some of them have failed the availability check.
+			//
+			// In this case, set failed and drifted placements; no diffed placements will be set (diffs
+			// will only occur when there is an apply failure or the report diff mode is on).
+			failedManifests := extractFailedResourcePlacementsFromWork(w)
+			failedResourcePlacements = append(failedResourcePlacements, failedManifests...)
+
+			driftedManifests := extractDriftedResourcePlacementsFromWork(w)
+			driftedResourcePlacements = append(driftedResourcePlacements, driftedManifests...)
+		default:
```

Comment:
isn't it better to explicitly have a case for that and then add an unexpected error in the default case so that we know we didn't miss anything.

Reply:
Got it; will send a subsequent PR.